subsys/bluetooth/shell/l2cap.c

```diff
@@ -0,0 +1,321 @@
+{
+
+	if (!shell_cmd_precheck(shell, argc >= 2, NULL, 0)) {
+		return;
+	}
+
+	if (server.psm) {
+		error(shell, "Already registered\n");
+		return;
+	}
+
+	server.psm = strtoul(argv[1], NULL, 16);
+
+	if (argc > 2) {
+		server.sec_level = strtoul(argv[2], NULL, 10);
+	}
+
+	if (bt_l2cap_server_register(&server) < 0) {
+		error(shell, "Unable to register psm\n");
+		server.psm = 0;
+	} else {
+		print(shell, "L2CAP psm %u sec_level %u registered\n",
+		      server.psm, server.sec_level);
+	}
+}
+
+static void cmd_connect(const struct shell *shell, size_t argc, char *argv[])
+{
+	u16_t psm;
+	int err;
+
+	if (!default_conn) {
+		error(shell, "Not connected\n");
+		return;
+	}
+
+	if (!shell_cmd_precheck(shell, argc == 2, NULL, 0)) {
+		return;
+	}
+
+	if (l2ch_chan.ch.chan.conn) {
+		error(shell, "Channel already in use\n");
+		return;
+	}
+
+	psm = strtoul(argv[1], NULL, 16);
+
+	err = bt_l2cap_chan_connect(default_conn, &l2ch_chan.ch.chan, psm);
+	if (err < 0) {
+		error(shell, "Unable to connect to psm %u (err %u)\n", psm,
+		      err);
+	} else {
+		print(shell, "L2CAP connection pending\n");
+	}
+}
+
+static void cmd_disconnect(const struct shell *shell, size_t argc, char *argv[])
+{
+	int err;
+
+	err = bt_l2cap_chan_disconnect(&l2ch_chan.ch.chan);
+	if (err) {
+		print(shell, "Unable to disconnect: %u\n", -err);
+	}
+}
+
+static void cmd_send(const struct shell *shell, size_t argc, char *argv[])
+{
+	static u8_t buf_data[DATA_MTU] = { [0 ... (DATA_MTU - 1)] = 0xff };
+	int ret, len, count = 1;
+	struct net_buf *buf;
+
+	if (argc > 1) {
+		count = strtoul(argv[1], NULL, 10);
+	}
+
+	len = min(l2ch_chan.ch.tx.mtu, DATA_MTU - BT_L2CAP_CHAN_SEND_RESERVE);
+
+	while (count--) {
+		buf = net_buf_alloc(&data_tx_pool, K_FOREVER);
+		net_buf_reserve(buf, BT_L2CAP_CHAN_SEND_RESERVE);
+
+		net_buf_add_mem(buf, buf_data, len);
+		ret = bt_l2cap_chan_send(&l2ch_chan.ch.chan, buf);
+		if (ret < 0) {
+			print(shell, "Unable to send: %d\n", -ret);
+			net_buf_unref(buf);
+			break;
+		}
+	}
+}
+
+static void cmd_recv(const struct shell *shell, size_t argc, char *argv[])
+{
+	if (argc > 1) {
+		l2cap_recv_delay = strtoul(argv[1], NULL, 10);
+	} else {
+		print(shell, "l2cap receive delay: %u ms\n",
+			     l2cap_recv_delay);
+	}
+}
+
+static void cmd_metrics(const struct shell *shell, size_t argc, char *argv[])
+{
+	const char *action;
+
+	if (argc < 2) {
+		print(shell, "l2cap rate: %u bps.\n", l2cap_rate);
+
+		return;
+	}
+
+	action = argv[1];
+
+	if (!strcmp(action, "on")) {
+		l2cap_ops.recv = l2cap_recv_metrics;
+	} else if (!strcmp(action, "off")) {
+		l2cap_ops.recv = l2cap_recv;
+	} else {
+		shell_help_print(shell, NULL, 0);
+		return;
+	}
+
+	print(shell, "l2cap metrics %s.\n", action);
+}
+
+#define HELP_NONE "[none]"
+
+SHELL_CREATE_STATIC_SUBCMD_SET(l2cap_cmds) {
+	SHELL_CMD(connect, NULL, "<psm>", cmd_connect),
+	SHELL_CMD(disconnect, NULL, HELP_NONE, cmd_disconnect),
+	SHELL_CMD(metrics, NULL, "<value on, off>", cmd_metrics),
```

Comment:
You can consider implementing subcommands: `off` and `on` without a handler. In such case body of function `cmd_metrics` remain unchanged but `off` and `on` can be prompted and autocompleated.

You can also implement handlers to `on` and `off`. Code will look like:

```
static void cmd_metrics_on(const struct shell *shell, size_t argc, char *argv[])
{
 	if (!shell_cmd_precheck(shell, argc == 1, NULL, 0)) {
		return;
	}

	l2cap_ops.recv = l2cap_recv_metrics;
}

static void cmd_metrics_off(const struct shell *shell, size_t argc, char *argv[])
{
 	if (!shell_cmd_precheck(shell, argc == 1, NULL, 0)) {
		return;
	}

	l2cap_ops.recv = l2cap_recv;
}

static void cmd_metrics(const struct shell *shell, size_t argc, char *argv[])
{
 	if (!shell_cmd_precheck(shell, argc == 1, NULL, 0)) {
		return;
	}
        print(shell, "Please specify valid argument.\r\n");
}

SHELL_CREATE_STATIC_SUBCMD_SET(sub_cmd_connect)
{
	SHELL_CMD(off, NULL, "switch off", cmd_metrics_off),
	SHELL_CMD(on, NULL, "switch on", cmd_metrics_on),
        SHELL_SUBCMD_SET_END
};

```

and modify this line:
`SHELL_CMD(metrics, NULL, "<value on, off>", cmd_metrics),`
to
`SHELL_CMD(metrics, &sub_cmd_connect, "switch value", cmd_metrics),`

This approach will limit strcmp calls and simplyfy commands usage.

Reply:
I didn't like the idea of going more than 2 subcommand, specially when all it would do is to show on/off to auto complete. Btw, I wish we didn't have to do the precheck ourselves so how about we define a format define parameters in the command description so the shell parse that to know how many parameters the command takes and do the precheck before calling the command handler. While at that we could also make it auto complete using the parameters not only commands (so we don't have to add command representing parameters).